src/_P009_MCP.ino

```diff
@@ -109,6 +109,22 @@ boolean Plugin_009(byte function, struct EventStruct *event, String& string)
         break;
       }
 
+    case PLUGIN_REQUEST:
+      {
+        String device = parseString(string, 1);
+        String command = parseString(string, 2);
+        String strPar1 = parseString(string, 3);
+        if (device == F("mcpgpio") && command == F("pinstate"))
```

Comment:
These String allocations make it well readable, but they keep using memory while in scope.

Since they are not used in multiple places, better use them to store in a boolean?
`bool deviceMatch = parseString(string, 1) == F("mcpgpio"));`
And maybe even better would be if we make another function `parseStringCompare(...)` which accepts a flashstringhelper and does not create a String object at all.

I guess we really should start looking into reducing all the String objects we're using. I think we're fragmenting the heap a lot and may have reached a point where memory allocation takes too long, or just fails silently, which leads to a crash.

Reply:
you mean something like that:
```
    case PLUGIN_REQUEST:
      {
        if (parseString(string, 1)==F("mcpgpio") && parseString(string, 2) == F("pinstate"))
        {
          int par1;
          if (validIntFromString(parseString(string, 3), par1)) {
            string = Plugin_009_Read(par1);
          }
          success = true;
        }
        break;
      }

```

Reply:
Yep, but since I liked the readability of your code, maybe storing it in a boolean keeps it readable and doesn't take much memory to store.
Only disadvantage of using booleans is that both compares and parse operations will be executed.
When using in the current code suggestion you now post, the second one will not be executed when the first returns false.

Reply:
I wrote the meaning of the parameters in the comments:
```
    case PLUGIN_REQUEST:
      {
        //parseString(string, 1) = device
        //parseString(string, 2) = command
        //parseString(string, 3) = gpio number
        if (parseString(string, 1) == F("mcpgpio") && parseString(string, 2) == F("pinstate"))
        {
          int par1;
          if (validIntFromString(parseString(string, 3), par1)) {
            string = Plugin_009_Read(par1);
          }
          success = true;
        }
        break;
      }

```

Reply:
in this case, code is clear and execution is faster :)

Reply:
pushing changes